src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -729,37 +729,45 @@ public BigInteger(String val) {
      * @see #bitLength()
      */
     public BigInteger(int numBits, Random rnd) {
-        byte[] magnitude = randomBits(numBits, rnd);
+        // randomBits() returns a zero length array if len == 0
+        this.mag = randomBits(numBits, rnd);
 
-        try {
-            // stripLeadingZeroBytes() returns a zero length array if len == 0
-            this.mag = stripLeadingZeroBytes(magnitude, 0, magnitude.length);
-
-            if (this.mag.length == 0) {
-                this.signum = 0;
-            } else {
-                this.signum = 1;
-            }
-            if (mag.length >= MAX_MAG_LENGTH) {
+        if (this.mag.length == 0) {
+            this.signum = 0;
+        } else {
+            this.signum = 1;
+            if (mag.length >= MAX_MAG_LENGTH)
                 checkRange();
-            }
-        } finally {
-            Arrays.fill(magnitude, (byte)0);
         }
     }
 
-    private static byte[] randomBits(int numBits, Random rnd) {
+    private static int[] randomBits(int numBits, Random rnd) {
         if (numBits < 0)
             throw new IllegalArgumentException("numBits must be non-negative");
-        int numBytes = (int)(((long)numBits+7)/8); // avoid overflow
-        byte[] randomBits = new byte[numBytes];
 
-        // Generate random bytes and mask out any excess bits
-        if (numBytes > 0) {
-            rnd.nextBytes(randomBits);
-            int excessBits = 8*numBytes - numBits;
-            randomBits[0] &= (byte)((1 << (8-excessBits)) - 1);
+        if (numBits == 0)
+            return new int[0];
+
+        final int numInts = ((numBits - 1) >> 5) + 1;
+        // numInts >= 1, since numBits > 0
+
+        int firstInt = rnd.nextInt() & (-1 >>> -numBits); // Mask out any excess bits
```

Comment:
Instead of `numBits`, shouldn't this use something like `excessBits` similar to line 760 in the old version?

Reply:
It's equivalent. If you watch the implementations of other methods in the class where a mask out is made, this method is often used.

Reply:
Indeed it looks as if that is correct.

Reply:
As specified in https://docs.oracle.com/javase/specs/jls/se21/html/jls-15.html#jls-15.19:
"If the promoted type of the left-hand operand is int, then only the five lowest-order bits of the right-hand operand are used as the shift distance. It is as if the right-hand operand were subjected to a bitwise logical AND operator & ([§15.22.1](https://docs.oracle.com/javase/specs/jls/se21/html/jls-15.html#jls-15.22.1)) with the mask value 0x1f (0b11111). The shift distance actually used is therefore always in the range 0 to 31, inclusive.
The value of n >>> s is n right-shifted s bit positions with zero-extension, where If n is negative and the type of the left-hand operand is int, then the result is equal to that of the expression (n >> s) + (2 << ~s). The added term (2 << ~s) cancels out the propagated sign bit.
Note that, because of the implicit masking of the right-hand operand of a shift operator, ~s as a shift distance is equivalent to 31-s when shifting an int value and to 63-s when shifting a long value."
So, since ~s == -s - 1, we have that -s  as a shift distance is equivalent to 32-s.
Since (excessBits == 32 * numInts - numBits), then (excessBits == (32 - (numBits & 0x1f)) & 0x1f), because (x & 0x1f == x mod 32). Since the five lowest-order bits of 32 are 0, then the five lowest-order bits of (32 - (numBits & 0x1f)) are the same of (-numBits), and so (excessBits == (-numBits) & 0x1f), therefore (-1 >>> excessBits == -1 >>> -numBits).

Reply:
Thanks for the proof.